File: toot/step_definitions/Product Filters.java
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.junit.Assert;
import java.time.Duration;

public class ProductFiltersStepDefinitions {
    WebDriver driver;
    WebDriverWait wait;
    ProductFiltersPage productFiltersPage;

    public ProductFiltersStepDefinitions(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        this.productFiltersPage = new ProductFiltersPage(driver);
    }

    public void givenUserIsOnProductSearchPage() {
        driver.get("http://example.com/products");
    }

    public void whenUserSelectsPriceRangeFilter(String priceRange) {
        productFiltersPage.selectPriceRange(priceRange);
    }

    public void thenSearchResultsShouldDisplayProductsWithinSelectedPriceRange() {
        Assert.assertTrue("Products not within selected price range", productFiltersPage.verifyPriceRangeFilter());
    }

    public void whenUserSelectsRatingFilter(String rating) {
        productFiltersPage.selectRatingFilter(rating);
    }

    public void thenSearchResultsShouldDisplayProductsWithSelectedRatingOrHigher() {
        Assert.assertTrue("Products not with selected rating or higher", productFiltersPage.verifyRatingFilter());
    }

    public void whenUserSelectsBothPriceRangeAndRatingFilters(String priceRange, String rating) {
        productFiltersPage.selectPriceRange(priceRange);
        productFiltersPage.selectRatingFilter(rating);
    }

    public void thenSearchResultsShouldDisplayProductsThatMeetBothCriteria() {
        Assert.assertTrue("Products not meeting both criteria", productFiltersPage.verifyCombinedFilters());
    }

    public void givenUserHasAppliedMultipleFilters() {
        productFiltersPage.applyMultipleFilters();
    }

    public void whenUserClearsAllFilters() {
        productFiltersPage.clearAllFilters();
    }

    public void thenSearchResultsShouldDisplayAllProductsWithoutAnyFiltersApplied() {
        Assert.assertTrue("Filters not cleared", productFiltersPage.verifyNoFiltersApplied());
    }

    public void whenUserAppliesFiltersThatMatchNoProducts() {
        productFiltersPage.applyNoMatchingFilters();
    }

    public void thenSearchResultsShouldDisplayMessageIndicatingNoProductsFound() {
        Assert.assertTrue("No products found message not displayed", productFiltersPage.verifyNoProductsFoundMessage());
    }

    public void givenUserHasAppliedFilters() {
        productFiltersPage.applyFilters();
    }

    public void whenUserRefreshesPage() {
        driver.navigate().refresh();
    }

    public void thenPreviouslySelectedFiltersShouldRemainApplied() {
        Assert.assertTrue("Filters not persisted after refresh", productFiltersPage.verifyFiltersPersisted());
    }
}

File: toot/step_definitions/Product Filters.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import ProductSearchPage from "../pages/productSearchPage";

Given("the user is on the product search page", () => {
  ProductSearchPage.visit();
});

When("the user selects a price range filter", () => {
  ProductSearchPage.selectPriceRangeFilter();
});

Then("the search results should display products within the selected price range", () => {
  ProductSearchPage.verifyProductsInPriceRange();
});

When("the user selects a rating filter", () => {
  ProductSearchPage.selectRatingFilter();
});

Then("the search results should display products with the selected rating or higher", () => {
  ProductSearchPage.verifyProductsWithRatingOrHigher();
});

When("the user selects both price range and rating filters", () => {
  ProductSearchPage.selectMultipleFilters();
});

Then("the search results should display products that meet both criteria", () => {
  ProductSearchPage.verifyProductsMeetingBothCriteria();
});

Given("the user has applied multiple filters", () => {
  ProductSearchPage.applyMultipleFilters();
});

When("the user clears all filters", () => {
  ProductSearchPage.clearAllFilters();
});

Then("the search results should display all products without any filters applied", () => {
  ProductSearchPage.verifyAllProductsDisplayed();
});

When("the user applies filters that match no products", () => {
  ProductSearchPage.applyNoMatchingFilters();
});

Then("the search results should display a message indicating no products were found", () => {
  ProductSearchPage.verifyNoProductsMessage();
});

Given("the user has applied filters", () => {
  ProductSearchPage.applyFilters();
});

When("the user refreshes the page", () => {
  ProductSearchPage.refreshPage();
});

Then("the previously selected filters should remain applied", () => {
  ProductSearchPage.verifyFiltersPersisted();
});